File: 2015/Day11.pl
:- use_module(library(clpz)).
:- use_module(library(lists)).
:- use_module(library(dcgs)).
:- use_module(library(time)).

not_ins(_, []).
not_ins(Xs, [Y|Ys]) :-
    maplist(#\=(Y), Xs),
    not_ins(Xs, Ys).

chars_codes(Chars, Codes) :-
    maplist(char_code, Chars, Codes).

char_code_domain(StartChar, EndChar, StartCode .. EndCode) :-
    char_code(StartChar, StartCode),
    char_code(EndChar, EndCode).

straight([]).
straight([_]).
straight([C1,C2|Cs]) :-
    C1 #= C2 + 1,
    straight([C2|Cs]).

after([C|Cs], [D|Ds]) :- 
    same_length(Cs, Ds), 
    C #< D.
after([C|Cs], [C|Ds]) :- 
    same_length(Cs, Ds), 
    after(Cs, Ds).

password(Codes) :-
    % passwords must be exactly eight lowercase letters
    length(Codes, 8),
    char_code_domain(a, z, A_to_Z),
    Codes ins A_to_Z,

    % passwords may not contain the letters i, o, or l
    chars_codes("iol", Excluded),
    not_ins(Codes, Excluded),
    reverse(Codes, CodesRev),

    % passwords must include one increasing straight of at least three letters
    length(Straight, 3),
    straight(Straight),
    phrase((...,seq(Straight),...), CodesRev),

    % passwords must contain at least two different, non-overlapping pairs
    phrase((...,seq([X,X]),...,seq([Y,Y]),...), CodesRev).

% FIXME: enumeration is a bit off. In the beginning solutions are repeated. Why?
password_between(Codes, LowerBound, UpperBound) :-
    password(Between),
    after(LowerBound, Between),
    after(Between, UpperBound),

    label(Between),

    ( password_between(Codes, LowerBound, Between) 
    ; Codes = Between
    ; password_between(Codes, Between, UpperBound)
    ).

solution(NextPassword) :-
    chars_codes("vzbxkghb", OldCodes),
    password_between(Codes, OldCodes, BoundCodes),
    chars_codes(NextPassword, Codes).

File: 2015/Day11.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where
import Data.SBV (Goal, sChar_, SChar, SString, EqSymbolic (sNotElem, (.==)), constrain, OrdSymbolic ((.<), (.>), (.>=)), sChars, sString, SymVal (literal), minimize, Metric (toMetricSpace, fromMetricSpace), SInteger, sChar, optimize, OptimizeStyle (Lexicographic), sat)
import Data.Functor ((<&>))
import Data.String (fromString)
import qualified Data.SBV.Char as SChar
import qualified Data.SBV.String as SString
import Data.List ( (\\) )
import Data.Foldable (for_)

goal :: String -> Goal
goal oldPass = do
    nextPass <- sString "password"

    -- passwords must be exactly eight lowercase letters
    -- passwords may not contain the letters i, o, or l
    let alphabet = literal $ ['a' .. 'z'] \\ "iol"
    constrain $ SString.length nextPass .== 8
    for_ [0,1,2,3,4,5,6,7] $ \i ->
        constrain $ (nextPass SString.!! i) `SChar.elem` alphabet

    -- new password must "come after" old password
    constrain $ literal oldPass .< nextPass
    constrain $ nextPass .< literal ""

    -- passwords must include one increasing straight of at least three letters
    charS1 <- sChar "straight char 1"
    charS2 <- sChar "straight char 2"
    charS3 <- sChar "straight char 3"
    let straight = SString.implode [charS1,charS2,charS3]
    constrain $ SChar.ord charS1 +1 .== SChar.ord charS2
    constrain $ SChar.ord charS2 +1 .== SChar.ord charS3
    constrain $ straight `SString.isInfixOf` nextPass

    -- passwords must contain at least two different, non-overlapping pairs
    charP1 <- sChar "first pair character"
    charP2 <- sChar "second pair character"

    let pair1 = SString.implode [charP1,charP1]
        pair2 = SString.implode [charP2,charP2]

        indexPair1 = SString.offsetIndexOf nextPass pair1 0
        indexPair2 = SString.offsetIndexOf nextPass pair2 (indexPair1+2)

    constrain $ indexPair1 .> -1
    constrain $ indexPair2 .> -1
    
    -- minimize "goal" (SString.strToNat nextPass)

main :: IO ()
main = do
    let oldPass = "vzbxkghb"
    putStrLn "Part 1: "

    let oldPass' = "vzbxxyzz"
    print =<< sat (goal oldPass')